Fix #584: Removed NestedScrollView from TopicPractice

Luffy18346 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
Fix #584: Removed NestedScrollView from TopicPractice 
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

Luffy18346 committed:
Removed Nestedscrollview from TopicPractice

Luffy18346 commented:
@rt4914 , I checked the mockup for TopicPracticeFragment and made some changes(mostly padding, margin) in UI of the screen.
Here is my updated screen


![Screenshot_2020-02-06-23-00-59-56_092c5771dc9bb06bcb5e5b6a9dadbf56 1](https://user-images.githubusercontent.com/39301561/73965572-7189ab80-493a-11ea-9368-883924f033e7.png)

Luffy18346 committed:
Reverted changes from .idea/codeStyles/Project.xml

Luffy18346 committed:
Removed SkillSelectionAdapter and redundant code from TopicPracticeFragmentPresenter and topic_practice_fragment layout file

Luffy18346 commented on app/src/main/java/org/oppia/app/topic/practice/TopicPracticeFragmentPresenter.kt:
```diff
@@ -4,98 +4,143 @@ import android.view.LayoutInflater
     }
-    getTopicPracticeViewModel().notifySelectedSkillList(selectedSkillIdList)
+    binding.topicPracticeSkillList.apply {
+      this.adapter?.notifyDataSetChanged()
```
The code `      this.adapter?.notifyDataSetChanged()
` is working fine when there are few items in the skill list but sometimes crashes or not performing correctly when the item count increased such that user has to scroll through the list.

This code gives error "Cannot call this method while RecyclerView is computing a layout or scrolling"  sometimes when user click on the checkbox and startScrolling. I have tried applying Runnable as a solution suggested in stackoverflow but then the Start button remain unclickable no matter how many items are selected.

Luffy18346 replied:
removed `notifyDataSetChanged ` call and directly used TopicPracticeFooterViewBinding to update the `isSubmitButtonActive` boolean. Issue resolved.

rt4914 replied:
I have mentioned some changes to presenter file. Maybe that can help.

Luffy18346 replied:
Ok I will look into it.

Luffy18346 commented:
@rt4914 , the testcases in the TopicPracticeFragmentTest are also need to change because of the changes in layout files. Can u suggest anything that I should do in the testcases?

Luffy18346 committed:
Solved RecyclerView SubmitButton issue

Luffy18346 committed:
removed unnecessary imports

rt4914 commented on app/src/main/java/org/oppia/app/topic/practice/TopicPracticeFragmentPresenter.kt:
```diff
@@ -5,97 +5,138 @@ import android.view.View
+  private fun bindFooterView(
+    binding: TopicPracticeFooterViewBinding,
+    model: TopicPracticeFooterViewModel
+  ) {
```
This can come to a single line:

`private fun bindFooterView(binding: TopicPracticeFooterViewBinding, model: TopicPracticeFooterViewModel) {`

Luffy18346 replied:
Ok, refactored the code.

rt4914 commented on app/src/main/java/org/oppia/app/topic/practice/TopicPracticeFragmentPresenter.kt:
```diff
@@ -5,97 +5,138 @@ import android.view.View
-      logger.e("TopicPracticeFragment", "Failed to retrieve topic", topic.getErrorOrNull()!!)
-    }
-    return topic.getOrDefault(Topic.getDefaultInstance())
+  private fun bindFooterView(
```
This can come to a single line: 
`private fun bindSkillView(binding: TopicPracticeSkillViewBinding, model: TopicPracticeSkillSummaryViewModel) {`

Luffy18346 replied:
Ok, refactored the code.

rt4914 commented on app/src/main/java/org/oppia/app/topic/practice/TopicPracticeViewModel.kt:
```diff
@@ -1,24 +1,73 @@
+
+  private fun processTopicPracticeSkillList(topic: Topic): List<TopicPracticeItemViewModel> {
+
+    // List starting with the header
```
remove this comment and also empty line above it.

Luffy18346 replied:
removed the comment and empty line above it.

rt4914 commented on app/src/main/java/org/oppia/app/topic/practice/TopicPracticeViewModel.kt:
```diff
@@ -1,24 +1,73 @@
+      TopicPracticeHeaderViewModel() as TopicPracticeItemViewModel
+    ))  
+
+    // Add the Skills in the list
```
remove this comment

Luffy18346 replied:
removed it

rt4914 commented on app/src/main/java/org/oppia/app/topic/practice/TopicPracticeViewModel.kt:
```diff
@@ -1,24 +1,73 @@
+      TopicPracticeSkillSummaryViewModel(skill) as TopicPracticeItemViewModel
+    })
+
+    // Add Footer in the list
```
Remove this comment

Luffy18346 replied:
removed it too.

rt4914 commented on app/src/main/java/org/oppia/app/topic/practice/practiceitemviewmodel/TopicPracticeFooterViewModel.kt:
```diff
@@ -0,0 +1,14 @@
+  fun onStartButtonClicked() {
+    routeToQuestionPlayerListener.routeToQuestionPlayer(ArrayList())
+  }
+
```
Remove this empty line

Luffy18346 replied:
removed it.

rt4914 commented on app/src/main/java/org/oppia/app/topic/practice/practiceitemviewmodel/TopicPracticeHeaderViewModel.kt:
```diff
@@ -0,0 +1,7 @@
+import org.oppia.app.topic.practice.TopicPracticeFragment
+
+/** Header view model for the recycler view in [TopicPracticeFragment]. */
+class TopicPracticeHeaderViewModel(
```
Write it in single line: `class TopicPracticeHeaderViewModel : TopicPracticeItemViewModel()`

Luffy18346 replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/app/topic/practice/practiceitemviewmodel/TopicPracticeItemViewModel.kt:
```diff
@@ -0,0 +1,7 @@
+import org.oppia.app.viewmodel.ObservableViewModel
+
+/** Super-class for generalising different views for the recyclerView in [TopicPracticeFragment] */
+abstract class TopicPracticeItemViewModel: ObservableViewModel()
```
Nit spacing issue fix: 
`abstract class TopicPracticeItemViewModel : ObservableViewModel()`

Luffy18346 replied:
Nit spacing issue fixed.

rt4914 commented on app/src/main/java/org/oppia/app/topic/practice/practiceitemviewmodel/TopicPracticeSkillSummaryViewModel.kt:
```diff
@@ -0,0 +1,9 @@
+/** Skill summary view model for the recycler view in [TopicPracticeFragment]. */
+class TopicPracticeSkillSummaryViewModel(
+  val skillSummary: SkillSummary
+) : TopicPracticeItemViewModel()
```
Write this in single line:

`class TopicPracticeSkillSummaryViewModel(val skillSummary: SkillSummary) : TopicPracticeItemViewModel()`

Luffy18346 replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/app/topic/practice/TopicPracticeFragmentPresenter.kt:
```diff
@@ -5,97 +5,138 @@ import android.view.View
       selectedSkillIdList.add(skillId)
     }
-    getTopicPracticeViewModel().notifySelectedSkillList(selectedSkillIdList)
+    topicPracticeFooterViewBinding.isSubmitButtonActive = selectedSkillIdList.isNotEmpty()
```
Update:
```
    if (::topicPracticeFooterViewBinding.isInitialized) {
      topicPracticeFooterViewBinding.isSubmitButtonActive = selectedSkillIdList.isNotEmpty()
    }
```

We need to make sure that the footer has been initialised before working on its properties.

Luffy18346 replied:
Ok, I have applied the changes. code is working well. As an alternative of above change,       `topicPracticeFooterViewBinding?.isSubmitButtonActive = selectedSkillIdList.isNotEmpty()` this can also work right?

rt4914 commented on app/src/main/java/org/oppia/app/topic/practice/TopicPracticeFragmentPresenter.kt:
```diff
@@ -5,97 +5,138 @@ import android.view.View
-
-  internal fun onStartButtonClicked() {
-    routeToQuestionPlayerListener.routeToQuestionPlayer(selectedSkillIdList)
+    topicPracticeFooterViewBinding.isSubmitButtonActive = selectedSkillIdList.isNotEmpty()
```
Update: Ditto as above
```
    if (::topicPracticeFooterViewBinding.isInitialized) {
      topicPracticeFooterViewBinding.isSubmitButtonActive = selectedSkillIdList.isNotEmpty()
    }
```

Luffy18346 replied:
Ok, I have applied the changes.

rt4914 reviewed (changes requested):
I have mentioned some changes. Add more number of skills to topic to test how does this code changes affect the app.

rt4914 commented:
@Luffy18346 Change the title to `Fix #584: Removed NestedScrollView from TopicPractice`

rt4914 commented:
> @rt4914 , the testcases in the TopicPracticeFragmentTest are also need to change because of the changes in layout files. Can u suggest anything that I should do in the testcases?

@Luffy18346 I think checking this current functionality is more important currently, once we are sure that the code is perfect then we can easily figure out how test cases should be implemented. Whatever error we are facing right now should be somehow tested in test cases too.

Luffy18346 commented:
> @Luffy18346 Change the title to `Fix #584: Removed NestedScrollView from TopicPractice`

@rt4914 Changed the title to `Fix #584: Removed NestedScrollView from TopicPractice`

Luffy18346 commented:
> > @rt4914 , the testcases in the TopicPracticeFragmentTest are also need to change because of the changes in layout files. Can u suggest anything that I should do in the testcases?
> 
> @Luffy18346 I think checking this current functionality is more important currently, once we are sure that the code is perfect then we can easily figure out how test cases should be implemented. Whatever error we are facing right now should be somehow tested in test cases too.

Ok, got it.

Luffy18346 committed:
Applied suggested changes in all the mentioned files.

Luffy18346 committed:
Merge remote-tracking branch 'upstream/develop' into lowfi-remove-nestedscrollview-from-topicpractice

rt4914 reviewed (changes requested):
LGTM. Now you can focus on test cases. For test-cases, just note that we want to achieve same functionality as current test cases, you will just need to update the code as per new implementation.

Also, as the entire screen is a recyclerview now, you will need to scroll to a specific item because checking/performing anything on that item.

Luffy18346 committed:
TopicPracticeFragment test-cases implementations changed.

Luffy18346 commented:
@rt4914 I have refactored the code in `TopicPracticeFragmentTest`. All test-cases are passing in emulator. 
I have also made changes in other files because earlier I was passing empty arraylist instead of selected skillList, inthe `QuestionPlayerActivity`. I have rectified that mistake.

Luffy18346 commented:
@rt4914 I wanted to ask whether in the skillList, every skill will have unique id? If yes, then I have checked everything and it's working fine.
Another thing I wanted to ask was that I have implemented almost everything in logic in the `TopicPracticeFragmentPresenter`. So, is this ok? or if it can be optimised then how?

rt4914 commented on app/src/main/java/org/oppia/app/topic/practice/TopicPracticeViewModel.kt:
```diff
@@ -1,24 +1,67 @@
+  private val topicController: TopicController
 ) : ViewModel() {
+  private val itemViewModelList: MutableList<TopicPracticeItemViewModel> = ArrayList()
+  private val routeToQuestionPlayerListener = activity as RouteToQuestionPlayerListener
```
This is not getting used anymore, remove this and its corresponding import.

Luffy18346 replied:
removed the line and corresponding import.

rt4914 commented on app/src/main/java/org/oppia/app/topic/practice/TopicPracticeViewModel.kt:
```diff
@@ -1,24 +1,67 @@
 import javax.inject.Inject
 
-/** [ViewModel] for showing skills in practice fragment. */
+/** The ViewModel for [TopicPracticeFragment]. */
```
Actually it should be `/** [ViewModel] for [TopicPracticeFragment]. */`

Luffy18346 replied:
Done.

rt4914 commented on app/src/main/java/org/oppia/app/topic/practice/practiceitemviewmodel/TopicPracticeFooterViewModel.kt:
```diff
@@ -0,0 +1,7 @@
+import org.oppia.app.topic.practice.TopicPracticeFragment
+
+/** Footer view model for the recycler view in [TopicPracticeFragment]. */
+class TopicPracticeFooterViewModel : TopicPracticeItemViewModel()
```
EOF is missing. There is should be one blank line at the end in every file.

Luffy18346 replied:
Done.

rt4914 commented on app/src/sharedTest/java/org/oppia/app/topic/practice/TopicPracticeFragmentTest.kt:
```diff
@@ -6,16 +6,19 @@ import android.content.res.Configuration
 import androidx.test.core.app.ActivityScenario
 import androidx.test.core.app.ApplicationProvider
 import androidx.test.espresso.Espresso.onView
+import androidx.test.espresso.action.ViewActions
```
Remove these unused imports in this file.

Luffy18346 replied:
Done

rt4914 reviewed (changes requested):
Great work on these test cases and the overall implementation. I have requested some final nit changes. PTAL

rt4914 commented:
> @rt4914 I have refactored the code in `TopicPracticeFragmentTest`. All test-cases are passing in emulator.
> I have also made changes in other files because earlier I was passing empty arraylist instead of selected skillList, inthe `QuestionPlayerActivity`. I have rectified that mistake.

Great. Thanks.

rt4914 commented:
> @rt4914 I wanted to ask whether in the skillList, every skill will have unique id? If yes, then I have checked everything and it's working fine.
> Another thing I wanted to ask was that I have implemented almost everything in logic in the `TopicPracticeFragmentPresenter`. So, is this ok? or if it can be optimised then how?

SkillIds are unique.
Login should go in presenter only, so your code is correct.

Luffy18346 committed:
Removed unused imports

Luffy18346 committed:
Merge remote-tracking branch 'upstream/develop' into lowfi-remove-nestedscrollview-from-topicpractice

rt4914 reviewed (approved):
LGTM, thanks.
